### PR TITLE
journal: delete old journal entries and blocks

### PR DESCRIPTION
**Pull request by strib**
Do you think this needs tests?

Issue: KBFS-1410

### CODE REVIEW

File: libkbfs/block_journal.go [hunk 1]
```diff
@@ -895,6 +901,16 @@ func (j *blockJournal) removeFlushedEntry(ctx context.Context,
 		}
 	}
 
+	// Remove the corresponding block from disk.  TODO: eventually we
```
Comment (akalin-keybase) [inline libkbfs/block_journal.go]: I think it's cleaner to leave the initial part of this function as is, and have something like this here:

```
if entry.Op == putOp || entry.Op == addRefOp {
  id, _, err := ...
  if err != nil {
    ...
  }

  if !j.hasRef(id) {
    ...
  }
}
```

**Comment by akalin-keybase** [comment]
Yeah, mind adding basic tests, also?

Also, should we clean up the MDs as we remove them also, or leave that for another PR? The only tricky thing is the `convertBranch` case.

**Comment by strib** [comment]
Added tests, MD garbage collection (forgot about this!) and simplified/fixed the block deletion code.  Ready for another look!

File: libkbfs/md_journal.go [hunk 2]
```diff
@@ -464,6 +464,17 @@ func (j *mdJournal) convertToBranch(
 	j.j = tempJournal
 	j.branchID = bid
 
+	// Garbage-collect the old non-branch entries.  TODO: we'll
```
Comment (akalin-keybase) [inline libkbfs/md_journal.go]: Actually, I think you don't need this after all. See line 462, and the referred-to defer (which nukes `journalTempDir`.)

Comment (akalin-keybase) [inline libkbfs/md_journal.go]: D'oh, never mind, I see it now.

File: libkbfs/tlf_journal_test.go [hunk 3]
```diff
@@ -548,6 +549,20 @@ func requireJournalEntryCounts(t *testing.T, j *tlfJournal,
 	require.Equal(t, expectedMDEntryCount, mdEntryCount)
 }
 
+func testMDJournalGCd(t *testing.T, j *mdJournal) {
```
Comment (akalin-keybase) [inline libkbfs/tlf_journal_test.go]: Should we do this in md_journal_test.go also (or instead)?

File: libkbfs/md_journal.go (hunk 2 above)
Comment (akalin-keybase) [inline libkbfs/md_journal.go]: I think we can do slightly better here, by cleaning up the branch entries if we encounter an error halfway. I think all you need is to keep track of the new md IDs, clean them up in the same defer that nukes the temp dir, and then set it to `allMdIDs` here.

I think one of the tests in `md_journal_test.go` has a case where the branch conversion fails, so if you add a GC check to those tests, we'll get coverage for this, too.

File: libkbfs/tlf_journal_test.go (hunk 3 above)
Comment (strib) [inline libkbfs/tlf_journal_test.go]: That's how I started doing it, but none of those tests fully remove all flushed entries at the end of the test and I didn't think it was important enough to add new tests for this.

File: libkbfs/md_journal.go (hunk 2 above)
Comment (strib) [inline libkbfs/md_journal.go]: Can you tell me which test?  I can't figure it out.

Comment (strib) [inline libkbfs/md_journal.go]: Oh nevermind, I see it.  The "Atomic" one.  Unfortunately it doesn't flush or remove flushed entries.

**Comment by strib** [comment]
@akalin-keybase: add those MD flush tests, and it indeed revealed some issues which I fixed as well.  Good call!

**Comment by strib** [comment]
(Ready for another look!)

File: libkbfs/md_journal.go (hunk 2 above)
Comment (akalin-keybase) [inline libkbfs/md_journal.go]: This comment probably belongs in the defer, too

**Comment by akalin-keybase** [comment]
lgtm after comment

**Pull request merged**